sycl/include/syclcompat/launch_policy.hpp

```diff
@@ -0,0 +1,238 @@
+/***************************************************************************
+ *
+ *  Copyright (C) Codeplay Software Ltd.
+ *
+ *  Part of the LLVM Project, under the Apache License v2.0 with LLVM
+ *  Exceptions. See https://llvm.org/LICENSE.txt for license information.
+ *  SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCL compatibility extension
+ *
+ *  launch.hpp
+ *
+ *  Description:
+ *    launch functionality for the SYCL compatibility extension
+ **************************************************************************/
+
+#pragma once
+
+#include "sycl/ext/oneapi/experimental/enqueue_functions.hpp"
+#include "sycl/ext/oneapi/properties/properties.hpp"
+#include <sycl/event.hpp>
+#include <sycl/nd_range.hpp>
+#include <sycl/queue.hpp>
+#include <sycl/range.hpp>
+
+#include <syclcompat/defs.hpp>
+#include <syclcompat/device.hpp>
+#include <syclcompat/dims.hpp>
+#include <syclcompat/traits.hpp>
+
+namespace syclcompat {
+namespace experimental {
+
+namespace sycl_exp = sycl::ext::oneapi::experimental;
+
+// Wrapper for kernel sycl_exp::properties
+template <typename Properties> struct kernel_properties {
+  static_assert(sycl_exp::is_property_list_v<Properties>);
+  using Props = Properties;
+
+  template <typename... Props>
+  kernel_properties(Props... properties) : props{properties...} {}
+
+  Properties props;
+};
+
+template <typename... Props>
+kernel_properties(Props... props)
+    -> kernel_properties<decltype(sycl_exp::properties(props...))>;
+
+// Wrapper for launch sycl_exp::properties
+template <typename Properties> struct launch_properties {
+  static_assert(sycl_exp::is_property_list_v<Properties>);
+  using Props = Properties;
+
+  template <typename... Props>
+  launch_properties(Props... properties) : props{properties...} {}
```

Comment:
This is more of a question, I am just curious from a design aspect, why did you not let the user pass a property list and you are constructing it internally instead ?

Reply:
Since the `launch_config` takes a bunch of different options (local mem, kernel properties, launch properties), I wanted to make the ctor variadic so the user can request only what they need. In order to do that in a way which is unambiguous, the `properties` need to be wrapped (in `compat::kernel_properties`, `compat_exp::launch_properties`). Otherwise even the compiler can't really tell whether a `properties` object is launch or kernel properties.

At present the wrappers take a variadic list of individual `property`s, as opposed to a `properties`. It *could* take a `properties` instead if that was better?